crates/revm/src/context/evm_context.rs

```diff
@@ -7,74 +8,55 @@ use crate::{
     interpreter::{
         return_ok, CallInputs, Contract, Gas, InstructionResult, Interpreter, InterpreterResult,
     },
-    primitives::{Address, Bytes, EVMError, Env, U256},
+    primitives::{
+        Address, Bytes, ChainSpec, EVMError, EVMErrorForChain, Env, TransactionValidation, U256,
+    },
     ContextPrecompiles, FrameOrResult, CALL_STACK_LIMIT,
 };
-use core::{
-    fmt,
-    ops::{Deref, DerefMut},
-};
+use core::ops::{Deref, DerefMut};
 use std::boxed::Box;
 
 /// EVM context that contains the inner EVM context and precompiles.
-pub struct EvmContext<DB: Database> {
+#[derive_where(Clone, Debug; ChainSpecT::Block, ChainSpecT::Transaction, DB, DB::Error)]
+pub struct EvmContext<ChainSpecT: ChainSpec, DB: Database> {
     /// Inner EVM context.
-    pub inner: InnerEvmContext<DB>,
+    pub inner: InnerEvmContext<ChainSpecT, DB>,
     /// Precompiles that are available for evm.
-    pub precompiles: ContextPrecompiles<DB>,
-}
-
-impl<DB: Database + Clone> Clone for EvmContext<DB>
-where
-    DB::Error: Clone,
-{
-    fn clone(&self) -> Self {
-        Self {
-            inner: self.inner.clone(),
-            precompiles: ContextPrecompiles::default(),
-        }
-    }
+    pub precompiles: ContextPrecompiles<ChainSpecT, DB>,
 }
 
-impl<DB> fmt::Debug for EvmContext<DB>
-where
-    DB: Database + fmt::Debug,
-    DB::Error: fmt::Debug,
-{
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.debug_struct("EvmContext")
-            .field("inner", &self.inner)
-            .field("precompiles", &self.inner)
```

Comment:
I used the `derive_where` procedural macro to avoid errors like this (double usage of `self.inner`).